components/suggest/src/yelp.rs

```diff
@@ -0,0 +1,322 @@
+const MAX_MODIFIER_WORDS_LENGTH: usize = 2;
+
+static URL_TEMPLATE: Mutex<String> = Mutex::new(String::new());
+
+/// Insert Yelp dataset into db.
+pub(crate) fn insert_dataset(
+    conn: &Connection,
+    scope: &SqlInterruptScope,
+    dataset: &DownloadedYelpDataset,
+) -> Result<()> {
+    for keyword in &dataset.subjects {
+        scope.err_if_interrupted()?;
+        conn.execute("INSERT INTO yelp_subjects(keyword) VALUES(?)", [keyword])?;
+    }
+
+    for keyword in &dataset.pre_modifiers {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_modifiers(type, keyword) VALUES(:type, :keyword)",
+            named_params! {
+                ":type": Modifier::PreModifier,
+                ":keyword": keyword,
+            },
+        )?;
+    }
+
+    for keyword in &dataset.post_modifiers {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_modifiers(type, keyword) VALUES(:type, :keyword)",
+            named_params! {
+                ":type": Modifier::PostModifier,
+                ":keyword": keyword,
+            },
+        )?;
+    }
+
+    for sign in &dataset.location_signs {
+        scope.err_if_interrupted()?;
+        conn.execute(
+            "INSERT INTO yelp_location_signs(keyword, need_location) VALUES(:keyword, :need_location)",
+            named_params! {
+                ":keyword": sign.keyword,
+                ":need_location": sign.need_location,
+            },
+        )?;
+    }
+
+    *URL_TEMPLATE.lock().unwrap() = dataset.template.to_string();
+
+    Ok(())
+}
+
+/// Fetch Yelp suggestion from given user's query.
+pub(crate) fn fetch_suggestion(
+    conn: &Connection,
+    query: &SuggestionQuery,
+) -> Result<Option<Suggestion>> {
+    if !query.providers.contains(&SuggestionProvider::Yelp) {
+        return Ok(None);
+    }
+
+    if query.keyword.len() > MAX_QUERY_LENGTH {
+        return Ok(None);
+    }
+
+    if !&query.keyword.contains(" ") {
+        // Try to create by one word.
+        return to_suggestion(conn, query, &query.keyword, None, None, None, None, false);
+    }
+
+    // Find the location sign and the location.
+    let (query_without_location, location_sign, location, need_location) =
+        find_location(conn, &query.keyword)?;
+
+    if let (Some(_), false) = (&location, need_location) {
+        // The location sign does not need the specific location, but user is setting something.
+        return Ok(None);
+    }
+
+    if query_without_location.is_empty() {
+        // No remained query.
+        return Ok(None);
+    }
+
+    // Find the modifiers.
+    let (subject_candidate, pre_modifier, post_modifier) =
+        find_modifiers(conn, &query_without_location)?;
+
+    to_suggestion(
+        conn,
+        query,
+        &subject_candidate,
+        pre_modifier,
+        post_modifier,
+        location_sign,
+        location,
+        need_location,
+    )
+}
+
+fn to_suggestion(
```

Comment:
Can you also add a docstring for this function describing what it does? An example would be even nicer.

Reply:
The number of arguments (and optionals) that this function takes definitely makes it a little tricky to read—and it also does a combination of I/O (via `is_subject`) and string manipulation.

It'll be a bit more code, but I'm wondering how you'd feel about:

1. Lifting the `is_subject` check into `fetch_suggestion`. That function already calls other methods that do I/O (like `find_location` and `find_modifiers`), and then this function can focus on just the string manipulation. This would also make it easier to unit test separately, without having to pull in a `Connection`.
2. Changing this function to either take a struct of options—or make it a method on that struct!

Something like:

```rs
struct SuggestionBuilder<'a> {
    query: &'a SuggestionQuery,
    subject: &'a str,
    pre_modifier: Option<String>,
    // ...
}

impl<'a> From<SuggestionBuilder<'a>> for Suggestion {
    fn from(builder: SuggestionBuilder<'a>) -> Suggestion {
        let mut description = String::from(builder.subject);
        
        if let Some(modifier) = &self.pre_modifier {
            // ...
        }

        Suggestion::Yelp {
            url,
            // Use user’s query as title as it is.
            title: builder.query.keyword.clone(),
        }
    }
}
```

Then you could use it like this to build your `Suggestion` from a `SuggestionBuilder`:

```rs
let builder = SuggestionBuilder {
    query: query,
    subject: &subject,
    // ...
};

Ok(Some(suggestion.into()))
```

How does that sound?

Reply:
Oh, cool! I will do that :)